Documentation for Workload Sets feature (#42109)
dotnet/docs: docs/core/tools/dotnet-workload-sets.md

---
title: .NET SDK workload sets
description: "Learn how to use workload sets to control which versions of workloads get installed, updated, or restored."
author: tdykstra
ms.author: tdykstra
ms.service: dotnet
ms.topic: how-to
ms.date: 08/02/2024

#customer intent: As a developer using optional workloads, I want to keep my workload versions in sync so that I can avoid conflicts betwween versions. I also want to avoid unexpected workload updates, so that I can avoid disruptions in development.
---
<!--https://aka.ms/patterns-feedback-->

# .NET SDK workload sets

> The *workload sets* feature provides a version number that represents a group of .NET SDK workloads. The [install](dotnet-workload-install.md), [update](dotnet-workload-update.md), and [restore](dotnet-workload-restore.md) commands use this number in *workload set update mode* to provide the following benefits:

Reviewer: ```suggestion
The *workload sets* feature provides a version number that represents a group of .NET SDK workloads. The [install](dotnet-workload-install.md), [update](dotnet-workload-update.md), and [restore](dotnet-workload-restore.md) commands use this number in *workload-set update mode* to provide the following benefits:
```

> * You get control of the cadence of change for installed workload versions. The alternative mode of operation without using workload sets is called *loose manifests update mode*. In this mode, workloads are updated automatically as new versions of individual workloads are released onto any configured NuGet feeds. In workload set update mode, workloads stay at a specific workload set version until you explicitly change that version. So workloads stay at the same manifest version, which is specified by the workload set.

Reviewer: ```suggestion
* You get control of the cadence of change for installed workload versions. The alternative mode of operation without using workload sets is called *loose manifests update mode*. In this mode, workloads are updated automatically as new versions of individual workloads are released onto any configured NuGet feeds. In workload-set update mode, workloads stay at a specific workload-set version until you explicitly change that version. So workloads stay at the same manifest version, which is specified by the workload set.
```

* You can install and update a combination of workload versions that ship at the same time and are known to work together.
* You can ensure that everyone on your team is always working on the same workload versions.
* You don't have to use a rollback file to specify what workload version you want to be on.

<!--Workload sets are published to nuget.org with each release of the .NET SDK, under the package ID `Microsoft.NET.Workloads.<feature band>`-->

Here are some ways you can use workload sets:

> * "Pin" the `install` command to a specific workload set version.

Reviewer: ```suggestion
* "Pin" the `install` command to a specific workload-set version.
```

> * Update installed workloads to the latest available workload set version.

Reviewer: ```suggestion
* Update installed workloads to the latest available workload-set version.
```

> * Install or update to a specified workload set version.

Reviewer: ```suggestion
* Install or update to a specified workload-set version.
```

> * Restore to a specific workload set version using global.json.

Reviewer: ```suggestion
* Restore to a specific workload-set version using global.json.
```

> * Check your current update mode and workload set version.

Reviewer: ```suggestion
* Check your current update mode and workload-set version.
```

And you can still choose to install or update to the latest version of each individual workload, ignoring workload sets.

## Prerequisites

* [.NET 8.0.400 SDK](https://dotnet.microsoft.com/en-us/download/dotnet/8.0) or later.

>   In 8.0.400 SDK, the `dotnet workload` commands are in workload set update mode only when it's explicitly selected. Starting in [.NET 9 SDK](https://dotnet.microsoft.com/en-us/download/dotnet/9.0), `dotnet workload` commands are in workload set update mode by default.

Reviewer: ```suggestion
  In 8.0.400 SDK, the `dotnet workload` commands are in workload-set update mode only when it's explicitly selected. Starting in [.NET 9 SDK](https://dotnet.microsoft.com/en-us/download/dotnet/9.0), `dotnet workload` commands are in workload-set update mode by default.
```

## "Pin" the install command
<!--If this heading changes, update the link in \docs\includes\cli-version.md-->

> A `dotnet workload install` command with the `--version` option "pins" the `install` command in workload set update mode with the specified workload set version. 

Reviewer: ```suggestion
A `dotnet workload install` command with the `--version` option "pins" the `install` command in workload-set update mode with the specified workload-set version.
```

<!--todo: install without --version in workload set update mode pins to latest workload set?-->
The command no longer automatically installs the newest workload based on loose manifests. For example:

1. ```dotnetcli
   dotnet workload install aspire --version 9.0.100-preview.7.24414.1
   ```

   When this command runs:

   * It gets the workload set with the specified version.
   * From the workload set, it gets the manifest version of the specified workload.
   * It installs the manifest version of the workload.

>    * It stays in workload set update mode when it's done.

Reviewer: ```suggestion
   * It stays in workload-set update mode when it's done.
```

1. ```dotnetcli
   dotnet workload install maui-ios
   ```

   This command installs the workload set version `9.0.100-preview.7.24414.1`, since the preceding `install` command example pinned that workload set.

> Using `--version` with either `install` or `update` pins `install` to the specified version, but `update` is only pinned to workload set update mode, not to a workload set version. If you then run `dotnet workload update` without the `--version` option, the `update` command:

Reviewer: ```suggestion
Using `--version` with either `install` or `update` pins `install` to the specified version, but `update` is only pinned to workload-set update mode, not to a workload-set version. If you then run `dotnet workload update` without the `--version` option, the `update` command:
```

> * Updates workloads to the latest available workload set version.

Reviewer: ```suggestion
* Updates workloads to the latest available workload-set version.
```

* "Unpins" the `install` command.

> * Stays in workload set update mode.

Reviewer: ```suggestion
* Stays in workload-set update mode.
```

## Update using latest workload set

> To update installed workloads to the latest workload set version available on the configured feeds, run the following commands:

Reviewer: ```suggestion
To update installed workloads to the latest workload-set version available on the configured feeds, run the following commands:
```

1. ```dotnetcli
   dotnet workload config --update-mode workload-set
   ```

>    The preceding command is necessary only if you are currently in manifests update mode. If that isn't known, [check the current update mode](#check-the-update-mode-and-version). Workload set update mode is the default starting in .NET 9 SDK.

Reviewer: ```suggestion
   The preceding command is necessary only if you are currently in manifests update mode. If you don't know, [check the current update mode](#check-the-update-mode-and-version). Workload-set update mode is the default starting in .NET 9 SDK.
```

1. ```dotnetcli
   dotnet workload update
   ```

>    This command updates workloads to the latest workload set version.

Reviewer: ```suggestion
   This command updates workloads to the latest workload-set version.
```

> ## Update to a workload set version

Reviewer: ```suggestion
## Update to a workload-set version
```

> To specify a workload set version to update to, use the `--version` option of the `update` command. For example:

Reviewer: ```suggestion
To specify a workload-set version to update to, use the `--version` option of the `update` command. For example:
```

```dotnetcli
dotnet workload update --version 8.0.400
```

> Workload set update mode will be selected if it wasn't already selected.

Reviewer: ```suggestion
Workload-set update mode will be selected if it wasn't already selected.
```

## Restore workloads using global.json

> Create a `global.json` file to specify the workload set version for the restore command for a repository. The format looks like this example:

Reviewer: ```suggestion
Create a `global.json` file to specify the workload-set version for the restore command for a repository. The format looks like this example:
```

---
<!--https://aka.ms/patterns-feedback-->

# .NET SDK workload sets

The *workload sets* feature provides a version number that represents a group of .NET SDK workloads. The [install](dotnet-workload-install.md), [update](dotnet-workload-update.md), and [restore](dotnet-workload-restore.md) commands use this number in *workload-set update mode* to provide the following benefits:

* You control the cadence of change for installed workload versions. The alternative mode of operation without using workload sets is called *loose manifests update mode*. In this mode, workloads are updated automatically as new versions of individual workloads are released onto any configured NuGet feeds. In workload set update mode, workloads stay at a specific workload set version until you explicitly change that version.
* You can install and update a combination of workload versions that ship at the same time and are known to work together.
* You can ensure that everyone on your team is always working on the same workload versions.

> * You don't have to use a rollback file to specify what workload version you want to be on.

Reviewer 2: Workload sets are opt-in in 8.0.4xx versions of the .NET SDK but will become opt-out in a future release (we hope 9.0.1xx but we can't commit to that so not sure how to phrase it).

Author: I cover this difference in the Prerequisites section. Is it certain enough that 9.0.1xx will be opt-out that we can say that, and go back and change it if it turns out to not be accurate? Or better to just say it's opt-in and not discuss the possibility of opt-out happening until it happens?

Reviewer 3: It's not yet certain that 9.0.100 will be opt-out, so what you have in the Prereqs seems fine to me. We can always clarify once we have a decision made.

<!--Workload sets are published to nuget.org with each release of the .NET SDK, under the package ID `Microsoft.NET.Workloads.<feature band>`-->

Here are some ways you can use workload sets:

* "Pin" the `install` command to a specific workload-set version.
* Update installed workloads to the latest available workload-set version.

> * Install or update to a specified workload-set version.

Reviewer 2: what's the difference between the first and third items in this list?

Author: The first is about installing, the third is about updating; I'll remove "Install or" at the start of the third bullet.

Author: ```suggestion
* Update to a specified workload-set version.
```

## "Pin" the install command

A `dotnet workload install` command with the `--version` option "pins" the `install` command in workload set update mode with the specified workload set version.
<!--todo: install without --version in workload set update mode pins to latest workload set?-->
The command no longer automatically installs the newest workload based on loose manifests. For example:

1. ```dotnetcli
   dotnet workload install aspire --version 9.0.100-preview.7.24414.1
   ```

   When this command runs:

   * It selects workload set update mode if that isn't already selected.
   * It gets the workload set that has the specified version.
   * From the workload set, it gets the manifest version of the specified workload.
   * It installs the manifest version of the workload.
   * It stays in workload-set update mode when it's done.

1. ```dotnetcli
   dotnet workload install maui-ios
   ```

   This command installs the workload set version `9.0.100-preview.7.24414.1`, since the preceding `install` command example pinned that workload set.

> Using `--version` with either `install` or `update` pins `install` to the specified version, but `update` is only pinned to workload-set update mode, not to a workload-set version. If you then run `dotnet workload update` without the `--version` option, the `update` command:

Reviewer 2: ```suggestion
Using `--version` with either `install` or `update` pins `install` to the specified version, but `update` is only configured for workload-set update mode, not to a specific workload-set version. If you then run `dotnet workload update` without the `--version` option, the `update` command:
```

* Updates workloads to the latest available workload-set version.
* "Unpins" the `install` command.
* Stays in workload-set update mode.

## Update using latest workload set

To update installed workloads to the latest workload-set version available on the configured feeds, run the following commands:

1. ```dotnetcli
   dotnet workload config --update-mode workload-set
   ```

   The preceding command is necessary only if you are currently in manifests update mode. If you don't know, [check the current update mode](#check-the-update-mode-and-version). Workload-set update mode is the default starting in .NET 9 SDK.

1. ```dotnetcli
   dotnet workload update
   ```

   In workload set update mode, this command updates workloads to the latest workload set version, unless you have specified the workload set version in global.json.

## Update to a workload-set version

> To specify a workload set version to update to when you're not specifying it in global.json, use the `--version` option of the `update` command. For example:

Reviewer 2: Where should we inform customers that the workload sets will typically align with the .Net sdk releases so if they are using an 8.0.4xx SDK, they'll want an 8.0.4xx workload set. BTW, there's a new feature coming that will let you search for workload set versions that are available.

Reviewer 3: I think there should be a 'choosing a workload set version' subsection in this doc where we could say this. Main idea would be 

* each SDK has a matching workload set version
* in the future we'll give you ways to list workload set versions to see what's available and what they contain

And then each command that has a `--version` could link to that section heading.

```dotnetcli
dotnet workload update --version 8.0.400
```

Workload-set update mode will be selected if it wasn't already selected.

## Use global.json for the workload set version

Create a `global.json` file to specify the workload set version for a repository. The format looks like this example:

```json
{
  "sdk": {
    "workloadVersion": "9.0.100-preview.7.24414.1"
  }
}
```

With the current directory in the same repository and the CLI in workload set update mode, the `install`, `update`, and `restore` commands install workloads for the specified workload set version. If you don't have a global.json file, and you're in workload set update mode, the `restore` command installs the workload set version that was established when you switched from manifests update mode to workload sets update mode.

> <!--todo:what about if it's SDK 9 so you defaulted to workload set update mode, and you never established a workload set version -- does it go for latest available then?-->

Reviewer 2: yes, it'll go for the latest available workload set.

Author: ```suggestion
If you're using SDK 9, you defaulted to workload set update mode, and you never established a workload-set version. In that case, `restore` installs the latest workload set available on the configured feeds.
```

> <!--todo:if you're in manifests update mode, you stay in that mode and it doesn't seem to do a workload set update, i.e., it seems to ignore global.json(?)-->

Reviewer 2: if you have a workload set in the global.json, it will treat you as in workload set mode even if you have not run the config command or --version. global.json overrides those.

> <!--todo:if you're in manifests update mode, you stay in that mode and it doesn't seem to do a workload set update, i.e., it seems to ignore global.json(?)-->

Author: ```suggestion

If you have a workload-set version in the global.json file, the workload commands are in workload-set mode even if you haven't run the `config` command or used `--version`. The global.json file overrides those.
```

> 
> If you specify the workload set version in global.json, you can't use the `--version` option to specify the workload set version. To use the `--version` option in that case, run the command outside of the path containing the global.json file.

Author: ```suggestion
To use the `--version` option in that case, run the command outside of the path containing the global.json file.
```

If you don't specify the workload set version in global.json, you can use the `--version` option with the `restore` comand. In that case, the `restore` command selects workload set update mode before it restores workloads to the specified workload set version.

In manifests update mode, `restore` installs or updates workloads to the latest version of each individual workload.

## Check the update mode and version

To see the current update mode, run the `config` command with the `--update-mode` option without an argument. The mode is either workload-set` or `manifests`. For example:

```dotnetcli
dotnet workload config --update-mode
```

```output
workload-set
```

> To see the current workload set version, run `dotnet workload --version`.  If a workload set is installed, you see a version such as 9.0.100-preview.7.24414.1 or 8.0.401. For example:

Reviewer 2: fyi, there is no 8.0.401 workload set. @baronfel should we have shipped an 8.0.401 that had no changes in it from 8.0.400 since we did a hotfix for that sdk?

Reviewer 2: fwiw, there will be an 8.0.402 workload set

Author: ```suggestion
To see the current workload set version, run `dotnet workload --version`.  If a workload set is installed, you see a version such as 9.0.100-preview.7.24414.1 or 8.0.402. For example:
```

* You can install and update a combination of workload versions that ship at the same time and are known to work together.
* You can ensure that everyone on your team is always working on the same workload versions.
* You don't have to use a rollback file to specify what workload version you want to be on.

Here are some ways you can use workload sets:

* "Pin" the `install` command to a specific workload-set version.
* Update installed workloads to the latest available workload-set version.
* Update to a specified workload-set version.
* Specify the workload set version in global.json.
* Check your current update mode and workload-set version.

And you can still choose to install, update, or restore to the latest version of each individual workload, ignoring workload sets.

## Prerequisites

* [.NET 8.0.400 SDK](https://dotnet.microsoft.com/en-us/download/dotnet/8.0) or later.

>   In 8.0.400 SDK, the `dotnet workload` commands are in workload-set update mode only when it's explicitly selected. Starting in [.NET 9 SDK](https://dotnet.microsoft.com/en-us/download/dotnet/9.0), `dotnet workload` commands are in workload-set update mode by default.

Reviewer 3: I don't think we can say that .NET 9 is opt-out for certain, so I think we should just say what we are certain of here: .NET 8.0.400 is opt-in.

## "Pin" the install command

A `dotnet workload install` command with the `--version` option "pins" the `install` command in workload set update mode with the specified workload set version.
<!--todo: install without --version in workload set update mode pins to latest workload set?-->
The command no longer automatically installs the newest workload based on loose manifests. For example:

1. ```dotnetcli
   dotnet workload install aspire --version 9.0.100-preview.7.24414.1
   ```

   When this command runs:

   * It selects workload set update mode if that isn't already selected.
   * It gets the workload set that has the specified version.
   * From the workload set, it gets the manifest version of the specified workload.
   * It installs the manifest version of the workload.
   * It stays in workload-set update mode when it's done.

1. ```dotnetcli
   dotnet workload install maui-ios
   ```

>    This command installs the workload set version `9.0.100-preview.7.24414.1`, since the preceding `install` command example pinned that workload set.

Reviewer 3: ```suggestion
   This command installs the the `maui-ios` workload using the workload version from the workload set version `9.0.100-preview.7.24414.1`, since the preceding `install` command example pinned that workload set.
```

I think it's worth clarifying where the version of the workload being installed are coming from a bit like you did in the `aspire` example above.

In manifests update mode, `restore` installs or updates workloads to the latest version of each individual workload.

## Check the update mode and version

To see the current update mode, run the `config` command with the `--update-mode` option without an argument. The mode is either workload-set` or `manifests`. For example:

```dotnetcli
dotnet workload config --update-mode
```

```output
workload-set
```

To see the current workload set version, run `dotnet workload --version`.  If a workload set is installed, you see a version such as 9.0.100-preview.7.24414.1 or 8.0.402. For example:

```dotnetcli
dotnet workload --version
```

```output
9.0.100-preview.7.24414.1
```

In manifests mode, or if the workload set version isn't established yet after switching to workload set update mode, you see a version in the form of `<feature band>-manifests.<hash>`. For example:

```dotnetcli
dotnet workload --version
```

```output
9.0.100-manifests.cf958b56
```

## Workload-set release cadence

> Workload sets are published to nuget.org with each release of the .NET SDK, under the package ID `Microsoft.NET.Workloads.<feature band>`. If, for example, you're using an 8.0.4xx SDK, the best workload set to choose is the same 8.0.4xx version of workload set.

Author: @marcpopMSFT @baronfel Is this paragraph accurate? Sometimes packages having names that fit this pattern are workload sets, and sometimes they aren't:

* Microsoft.NET.Workloads.8.0.400 is a package and you can install workload set version 8.0.400
* Microsoft.NET.Workloads.9.0.100-preview.7 is a package, but if you try to install 9.0.100-preview.7 or 9.100.0-preview.7.24414.1 (latest version) you get a not found message

In the absence of the future CLI workload set versions list command, how can people know what workload set versions are available?

Reviewer 2: For a stable version of the SDK, we plan on always having a matching version. So 8.0.400 sdk can install 8.0.400 workload set, 401 can install a 401 set.

As you've noticed, previews are harder. One think we've done is updated the readme in the package to have the workload set info. So for the preview 7, we include that here: https://www.nuget.org/packages/Microsoft.NET.Workloads.9.0.100-preview.7/9.100.0-preview.7.24414.1#readme-body-tab

and now that you've made me look, even that is wrong and I need to go fix it as it has an extra .0. So that'll be fixed in future workload set releases but broken for 8.0.400 and 9.0.100-preview.7
